rts/motoko-rts/src/memory/ic.rs

```diff
@@ -72,29 +72,29 @@ impl Memory for IcMemory {
     unsafe fn alloc_words(&mut self, n: Words<u32>) -> Value {
         let bytes = n.to_bytes();
         // Update ALLOCATED
-        ALLOCATED += Bytes(u64::from(bytes.as_u32()));
+        let delta = u64::from(bytes.as_u32());
+        ALLOCATED += Bytes(delta);
 
         // Update heap pointer
-        let old_hp = HP;
-        let (new_hp, overflow) = old_hp.overflowing_add(bytes.as_u32());
-        // Check for overflow
-        if overflow {
-            rts_trap_with("Out of memory");
-        }
-        HP = new_hp;
+        let old_hp = u64::from(HP);
+        let new_hp = old_hp + delta;
 
         // Grow memory if needed
-        grow_memory(new_hp as usize);
+        grow_memory(new_hp);
+
+        debug_assert!(new_hp <= u64::from(core::u32::MAX));
+        HP = new_hp as u32;
 
         Value::from_ptr(old_hp as usize)
     }
 }
 
 /// Page allocation. Ensures that the memory up to, but excluding, the given pointer is allocated.
 #[inline(never)]
-unsafe fn grow_memory(ptr: usize) {
+unsafe fn grow_memory(ptr: u64) {
+    debug_assert!(ptr <= 2 * u64::from(core::u32::MAX));
```

Comment:
We can use the same condition as in the assert in line 85: `ptr <= u64::from(core::u32::MAX)` (i.e. compare against 4GiB, not 8GiB)

Reply:
No, I'm deliberately comparing against 8GB because I want to test the overflow detection later on (otherwise we will get different faults on debug and release builds).